packages/babel-plugin-proposal-partial-application/src/index.js

```diff
@@ -0,0 +1,151 @@
+          init.push(
+            t.assignmentExpression(
+              "=",
+              t.cloneNode(id),
+              t.arrayExpression([t.spreadElement(node.arguments[i].argument)]),
+            ),
+          );
+          node.arguments[i].argument = t.cloneNode(id);
+        } else {
+          init.push(
+            t.assignmentExpression("=", t.cloneNode(id), node.arguments[i]),
+          );
+          node.arguments[i] = t.cloneNode(id);
+        }
+      }
+    }
+    return init;
+  }
+
+  function replacePlaceholders(node, scope) {
+    const placeholders = [];
+    const args = [];
+
+    node.arguments.forEach(arg => {
+      if (t.isArgumentPlaceholder(arg)) {
+        const id = scope.generateUid("_argPlaceholder");
+        placeholders.push(t.identifier(id));
+        args.push(t.identifier(id));
+      } else {
+        args.push(arg);
+      }
+    });
+    return [placeholders, args];
+  }
+
+  return {
+    name: "proposal-partial-application",
+    inherits: syntaxPartialApplication,
+
+    visitor: {
+      CallExpression(path) {
+        if (!hasArgumentPlaceholder(path.node)) {
+          return;
+        }
+        const { node, scope } = path;
+        const functionLVal = path.scope.generateUidIdentifierBasedOnNode(
+          node.callee,
+        );
+
+        const argsInitializers = unwrapArguments(node, scope);
+        const [placeholdersParams, args] = replacePlaceholders(node, scope);
+
+        scope.push({ id: functionLVal });
+
+        if (node.callee.type === "MemberExpression") {
+          const receiverLVal = path.scope.generateUidIdentifierBasedOnNode(
+            node.callee.object,
+          );
+          scope.push({ id: receiverLVal });
+          const finalExpression = t.sequenceExpression([
+            t.assignmentExpression(
+              "=",
+              t.cloneNode(receiverLVal),
+              node.callee.object,
+            ),
+            t.assignmentExpression(
+              "=",
+              t.cloneNode(functionLVal),
+              t.memberExpression(
+                receiverLVal,
+                node.callee.property,
+                false,
+                false,
+              ),
+            ),
+            ...argsInitializers,
+            t.functionExpression(
+              node.callee.property,
+              placeholdersParams,
+              t.blockStatement(
+                [
+                  t.returnStatement(
+                    t.callExpression(
+                      t.memberExpression(
+                        functionLVal,
+                        t.identifier("call"),
+                        false,
+                        false,
+                      ),
+                      [receiverLVal, ...args],
+                    ),
+                  ),
+                ],
+                [],
+              ),
+              false,
+              false,
+            ),
+          ]);
+          path.replaceWith(finalExpression);
```

Comment:
Not big deal by any means, but since both of these end with replacing the path with a `sequenceExpression`, we could do:

```js
const sequenceParts = [];

if (node.callee.type === "MemberExpression") {
  sequenceParts.push(
    /* stuff */
  );
} else {
  sequenceParts.push(
    /* stuff */
  );
}

path.replaceWith(t.sequenceExpression(sequenceParts));
```

Reply:
good idea! made the necessary changes.